meta_client/src/types.rs

```diff
@@ -162,18 +162,58 @@ pub struct NodeInfo {
     pub shard_infos: Vec<ShardInfo>,
 }
 
+/// The status changes of a shard as following:
+///
+///```plaintext
+///   ┌────┐
+///   │Init│
+///   └──┬─┘
+///   ___▽___
+///  ╱       ╲     ┌─────┐
+/// ╱ Opening ╲____│Ready│
+/// ╲         ╱yes └──┬──┘
+///  ╲_______╱    ┌───▽──┐
+///               │Frozen│
+///               └──────┘
+/// ```
+/// When an open request comes in, shard can only be opened when it's in
+/// - `Init`, which means it has not been opened before.
+/// - `Opening`, which means it has been opened before, but failed.
+#[derive(Debug, Default, Clone, PartialEq, Eq, Serialize)]
+pub enum ShardStatus {
+    /// Created, but not opened
+    #[default]
+    Init,
+    /// In opening
+    Opening,
+    /// Healthy
+    Ready,
+    /// Further updates are prohibited
+    Frozen,
+}
+
 #[derive(Debug, Default, Clone, PartialEq, Eq, Serialize)]
 pub struct ShardInfo {
     pub id: ShardId,
     pub role: ShardRole,
     pub version: ShardVersion,
+    // This status is only used for request ceresdb send to ceresmeta via heartbeat
+    // When ceresdb receive this via open shard request, this field is meanless.
+    // TODO: Use different request and response body between ceresdb and
+    // ceresmeta.
+    pub status: ShardStatus,
 }
 
 impl ShardInfo {
     #[inline]
     pub fn is_leader(&self) -> bool {
         self.role == ShardRole::Leader
     }
+
+    #[inline]
+    pub fn is_opened(&self) -> bool {
+        matches!(self.status, ShardStatus::Ready | ShardStatus::Frozen)
+    }
```

Comment:
I think the logic is a bit strange.
+ `ShardInfo` is defined in meta_client.
+ However, it can only be modify in `cluster`.
+ And it is read in `meta_client` from `cluster`.

How about define a struct such as `ShardState` in `cluster` and use it?
And we convert it to `ShardInfo` when call `all_shards`, during the convert, we map `ShardStatus` inner `ceresdb` to which sent to `ceresmeta`?
And the strange calling of `is_opened` in pb encode can be removed.